### ISSUE TITLE
[BUG] - one volume belongs to different collections

### ISSUE THREAD

**Issue by kaaaq** [chrislusf/seaweedfs#2823]
Hi @chrislusf,

Based on [this issue](https://github.com/chrislusf/seaweedfs/issues/2754), I seek for finding the problem and found out masters are returning 6 different locations for one VolumeId with "replica 010",
This is the result:

1.
`http://weed-master-1.local:9333/dir/lookup?volumeId=6123,7a4fd80f542df670`
```

{

    "volumeOrFileId": "6123",
    "locations": [
        {
            "url": "weed-volume-028.local:8085",
            "publicUrl": "weed-volume-028.local:8085"
        },
        {
            "url": "weed-volume-029.local:8080",
            "publicUrl": "weed-volume-029.local:8080"
        }
    ]

}

```

2.
`http://weed-master-2.local:9333/dir/lookup?volumeId=6123,7a4fd80f542df670`

```

{

    "volumeOrFileId": "6123",
    "locations": [
        {
            "url": "weed-volume-026.local:8084",
            "publicUrl": "weed-volume-026.local:8084"
        },
        {
            "url": "weed-volume-028.local:8085",
            "publicUrl": "weed-volume-028.local:8085"
        },
        {
            "url": "weed-volume-030.local:8082",
            "publicUrl": "weed-volume-030.local:8082"
        },
        {
            "url": "weed-volume-029.local:8080",
            "publicUrl": "weed-volume-029.local:8080"
        },
        {
            "url": "weed-volume-015.local:8080",
            "publicUrl": "weed-volume-015.local:8080"
        },
        {
            "url": "weed-volume-020.local:8085",
            "publicUrl": "weed-volume-020.local:8085"
        }
    ]

}

```

3.
`http://weed-master-3.local:9333/dir/lookup?volumeId=6123,7a4fd80f542df670`

```
{

    "volumeOrFileId": "6123",
    "locations": [
        {
            "url": "weed-volume-015.local:8080",
            "publicUrl": "weed-volume-015.local:8080"
        },
        {
            "url": "weed-volume-028.local:8085",
            "publicUrl": "weed-volume-028.local:8085"
        },
        {
            "url": "weed-volume-030.local:8082",
            "publicUrl": "weed-volume-030.local:8082"
        },
        {
            "url": "weed-volume-020.local:8085",
            "publicUrl": "weed-volume-020.local:8085"
        },
        {
            "url": "weed-volume-026.local:8084",
            "publicUrl": "weed-volume-026.local:8084"
        },
        {
            "url": "weed-volume-029.local:8080",
            "publicUrl": "weed-volume-029.local:8080"
        }
    ]

}

```
as you see, all masters are returning 6 unique volume servers:

```

weed-volume-015.local:8080
weed-volume-020.local:8085 
weed-volume-026.local:8084 
weed-volume-028.local:8085
weed-volume-029.local:8080
weed-volume-030.local:8082

```

So I browse to each volume UI and I saw that volumeID 6123 is for different collections at the same time:

1.

`http://weed-volume-015.local:8080/ui/index.html`

```
6123 	UserTest
```

2.

`http://weed-volume-020.local:8085/ui/index.html`

```
6123 	Cloud
```

3.

`http://weed-volume-026.local:8084 /ui/index.html`

```
6123 	WebTest
```

4.

`http://weed-volume-028.local:8085/ui/index.html`

```
6123 	UserTest
```

5.

`http://weed-volume-029.local:8080/ui/index.html`

```
6123 	Cloud
```

6.

`http://weed-volume-030.local:8082/ui/index.html`

```
6123 	WebTest
```

Why does one volume belong to 3 collections and not one?

This will cause many problems, e.g: I want to upload a file that is associated with "Cloud" collection, weed will return VolumeID 6123 and even for files that are associated with "UserTest" collection, weed will return volume 6123;
so for downloading files masters will return different and incorrect results (as you saw above) that will cause a  "404 error"

**Comment by kmlebedev** [comment]
@kaaaq The problem seems to be that before the collection was created, the master server lost the maxVolumeId values and created the collection with the already existing volume id. This happens when all the master servers were turned off, and then turned on without the resumeState true option

https://github.com/chrislusf/seaweedfs/discussions/2822

**Comment by Stitch10925** [comment]
Why would something as important as "resumeState" not be true as default? Or can you just restart the master with resumeState set to true to fix OP's issue?

**Comment by chrislusf** [comment]
> This happens when all the master servers were turned off, and then turned on without the resumeState true option

I am not sure whether the "resumeState" matters here or not. If the masters are configured properly, the raft protocol would ensure consistency and only one volume id would be assigned. "resumeState" just make it a bit faster.

**Comment by kmlebedev** [comment]
> > This happens when all the master servers were turned off, and then turned on without the resumeState true option
> 
> I am not sure whether the "resumeState" matters here or not. If the masters are configured properly, the raft protocol would ensure consistency and only one volume id would be assigned. "resumeState" just make it a bit faster.

The problem is that this is slow, since the grow function starts working earlier and creates new volume ids before all the hearbeats from the volume servers have arrived

**Comment by chrislusf** [comment]
Right. We can remove the "resumeState" option and always try to resume.

**Comment by kmlebedev** [comment]
> Right. We can remove the "resumeState" option and always try to resume.

It can also exclude the launch of the grow function until a collection with maxvolume appears in the topology

**Comment by kaaaq** [comment]
@chrislusf, @kmlebedev 

1. One suggestion could be using GUID instead of increasing VolumeId based on an integer number
Or better solution could be using the combination of Timestamp and integer number that increases sequentially 
I think the second way is better and more trustworthy than the first one
In this situation even if masters don't know about the last MaxVolumeId there wouldn't happen any conflict with volumes grow 

2. But in anyway, masters should know the last state of a cluster caused after restarting the cluster (all masters together),
it takes time to volumes join the master and send heartbeat to the leader, so in this time if any client requested for specific collection or volume, the master thinks there aren't enough volumes so it orders for creating a new volume,
This causes the creation of a lot of volumes and filling the capacity of max volumes that could be created in one volume server 
So masters should have metadata of the whole cluster.

**Comment by kmlebedev** [comment]
@chrislusf It seems like a raft server on the master, in addition to volumemaxid, stores the topology, but for some reason, when the leader changes, the topology is not saved.  Is this a bug or am I wrong?

**Comment by chrislusf** [comment]
@kmlebedev the topology is only dynamically collected and not stored.

**Comment by Woellchen** [comment]
what about the GUID idea from @kaaaq? it seems to be a better approach to avoid these kind of issues altogether. if the masters are supposed to stay kinda stateless, then simple increasing integers are probably not enough
